src/RDS.CaraBus.RabbitMQ/CaraBus.cs

```diff
@@ -197,7 +246,83 @@ public void CloseAndDisposeConnection()
             {
                 _connection.Close();
             }
+
             _connection.Dispose();
         }
+
+        private string GetTypeName(Type type)
+        {
+            if (type.IsConstructedGenericType)
+            {
+                var arguments = type.GetGenericArguments().Select(a => a.Name);
+                var formattedArguments = string.Join(", ", arguments);
+
+                var namespacePrefix = string.Empty;
+                if (type.DeclaringType != null)
+                {
+                    namespacePrefix = $"{type.DeclaringType.Name}";
+                }
+
+                return $"{namespacePrefix}+{type.Name}[{formattedArguments}]|{GetHash(type)}";
+            }
+            else
+            {
+                return $"{type.Name}|{GetHash(type)}";
+            }
+        }
+
+        private string GetHash(Type type)
+        {
+            using (var algorithm = SHA256.Create())
+            {
+                var hash = algorithm.ComputeHash(Encoding.ASCII.GetBytes(type.AssemblyQualifiedName));
+
+                string hashString = string.Empty;
+                foreach (byte x in hash)
+                {
+                    hashString += String.Format("{0:x2}", x);
+                }
+
+                return hashString;
+            }
+        }
+
+
+        private string GetExchangeName(string scope, Type type)
+        {
+            const int maxExchangeNameLength = 255;
+
+            var typeName = GetTypeName(type);
+
+            var exchangeName = $"{scope}|{typeName}";
+
+            if (exchangeName.Length > maxExchangeNameLength)
+            {
+                var extraQueueNameLength = exchangeName.Length - maxExchangeNameLength;
+                var lengthToRemove = extraQueueNameLength <= typeName.Length ? extraQueueNameLength : typeName.Length;
+                exchangeName = $"{scope}|{typeName.Remove(0, lengthToRemove)}";
+            }
+
+            return _messageExchanges.GetOrAdd($"{type.AssemblyQualifiedName}:{scope}", exchangeName);
```

Comment:
- method has side effect
- we should not calculate exchange name each method call. only if _messageExchanges has no specified key